src/libsyntax/ext/expand.rs

```diff
@@ -672,34 +746,46 @@ impl<'a, 'b> InvocationCollector<'a, 'b> {
         self.collect(kind, InvocationKind::Bang { mac: mac, ident: None, span: span })
     }
 
-    fn collect_attr(&mut self, attr: ast::Attribute, item: Annotatable, kind: ExpansionKind)
+    fn collect_attr(&mut self,
+                    attr: Option<ast::Attribute>,
+                    traits: Vec<(Symbol, Span)>,
+                    item: Annotatable,
+                    kind: ExpansionKind)
                     -> Expansion {
-        let invoc_kind = if attr.name() == "derive" {
-            if kind == ExpansionKind::TraitItems || kind == ExpansionKind::ImplItems {
-                self.cx.span_err(attr.span, "`derive` can be only be applied to items");
-                return kind.expect_from_annotatables(::std::iter::once(item));
-            }
-            InvocationKind::Derive { attr: attr, item: item }
-        } else {
-            InvocationKind::Attr { attr: attr, item: item }
-        };
-
-        self.collect(kind, invoc_kind)
+        if !traits.is_empty() &&
+           (kind == ExpansionKind::TraitItems || kind == ExpansionKind::ImplItems) {
+            self.cx.span_err(traits[0].1, "`derive` can be only be applied to items");
+            return kind.expect_from_annotatables(::std::iter::once(item));
+        }
+        self.collect(kind, InvocationKind::Attr { attr: attr, traits: traits, item: item })
     }
 
     // If `item` is an attr invocation, remove and return the macro attribute.
-    fn classify_item<T: HasAttrs>(&mut self, mut item: T) -> (T, Option<ast::Attribute>) {
-        let mut attr = None;
+    fn classify_item<T>(&mut self, mut item: T) -> (Option<ast::Attribute>, Vec<(Symbol, Span)>, T)
+        where T: HasAttrs,
+    {
+        let (mut attr, mut traits) = (None, Vec::new());
 
         item = item.map_attrs(|mut attrs| {
-            attr = self.cx.resolver.find_attr_invoc(&mut attrs).or_else(|| {
-                find_derive_attr(self.cx, &mut attrs)
-            });
+            if let Some(legacy_attr_invoc) = self.cx.resolver.find_legacy_attr_invoc(&mut attrs) {
+                attr = Some(legacy_attr_invoc);
+                return attrs;
+            }
+
+            traits = collect_derives(&mut self.cx, &attrs);
```

Comment:
No -- we don't call `classify_item` again after we collect a potential `proc_macro_attribute` invocation.
If the potential `proc_macro_attribute` invocation ends up being an inert attribute, we just mutate the invocation and re-resolve it (separate from this code-path).